File: frontend/src/types/compliance.ts
export type ComplianceStatus =
  | 'PASS'
  | 'FAIL'
  | 'REVIEW'
  | 'NOT_APPLICABLE'
  | 'PENDING'
  | 'BLOCKED';

export const ComplianceStatus = {
  PASS: 'PASS' as const,
  FAIL: 'FAIL' as const,
  REVIEW: 'REVIEW' as const,
  NOT_APPLICABLE: 'NOT_APPLICABLE' as const,
  PENDING: 'PENDING' as const,
  BLOCKED: 'BLOCKED' as const,
};

export interface ComplianceReviewItemResponse {
  requirement_code: string;
  requirement_name: string;
  category: string;
  review_type?: string | null;
  review_reason: string;
  source_name?: string | null;
  is_mandatory: boolean;
  is_critical: boolean;
}

export interface ComplianceResultItem {
  id: string;
  bid_id: string;
  tender_id: string;
  tender_requirement_id: string;
  requirement_code: string;
  requirement_name: string;
  category: string;
  requirement_type: string;
  compliance_status: ComplianceStatus;
  actual_value?: any;
  expected_value?: any;
  operator?: string;
  reason?: string;
  evidence?: Record<string, any>;
  source_verification_ids?: string[];
  is_mandatory: boolean;
  is_critical?: boolean;
  critical_failure?: boolean;
  weight?: number;
  evaluation_version: number;
  is_current: boolean;
  rule_version_id?: string | null;
  rule_version_number?: number | null;
  rule_version_code?: string | null;
  rule_version_name?: string | null;
  rule_version_operator?: string | null;
  rule_version_expected_value?: any;
  evaluated_at?: string;
  created_at: string;
  updated_at: string;
}

export interface ComplianceSummaryCounts {
  total: number;
  passed: number;
  failed: number;
  review: number;
  pending: number;
  not_applicable: number;
  blocked: number;
  mandatory_failures?: number;
  critical_failures?: number;
}

export interface BidComplianceSummaryResponse {
  bid_id: string;
  tender_id: string;
  tender_number?: string;
  bidder_name?: string;
  compliance_evaluation_complete: boolean;
  counts: ComplianceSummaryCounts;
  results: ComplianceResultItem[];
  review_items?: ComplianceReviewItemResponse[];
  evaluated_at?: string;
  evaluation_version?: number;
}

File: backend/app/schemas/compliance.py
"""
Compliance Engine Pydantic Schemas for Part 6A
Defines serialization models for compliance evaluation requests, rule results, and bid-level summaries.
"""

from datetime import datetime, timezone
from decimal import Decimal
from typing import Any, Dict, List, Optional
import uuid
from pydantic import BaseModel, ConfigDict, Field


class ComplianceReviewItemResponse(BaseModel):
    """Schema representing an item requiring human review."""
    model_config = ConfigDict(from_attributes=True)

    requirement_id: uuid.UUID
    requirement_code: str
    requirement_name: str
    category: str
    compliance_status: str
    review_type: Optional[str] = None
    reason: Optional[str] = None
    evidence: Optional[Dict[str, Any]] = None
    source_name: Optional[str] = None
    is_mandatory: bool = True
    is_critical: bool = False


class ComplianceResultItemResponse(BaseModel):
    """Schema representing an individual evaluated TenderRequirement result."""
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    bid_id: uuid.UUID
    tender_id: uuid.UUID
    tender_requirement_id: uuid.UUID
    requirement_code: str
    requirement_name: str
    category: str
    requirement_type: str
    compliance_status: str
    actual_value: Optional[Any] = None
    expected_value: Optional[Any] = None
    operator: Optional[str] = None
    reason: Optional[str] = None
    evidence: Optional[Dict[str, Any]] = None
    source_verification_ids: Optional[List[str]] = None
    document_id: Optional[uuid.UUID] = None
    document_name: Optional[str] = None
    page_number: Optional[int] = None
    download_url: Optional[str] = None
    evidence_snippet: Optional[str] = None
    confidence: Optional[float] = None
    verification_source: Optional[str] = None
    is_mandatory: bool = True
    is_critical: bool = False
    critical_failure: bool = False
    weight: Optional[Decimal] = None
    rule_version_id: Optional[uuid.UUID] = None
    rule_version_number: Optional[int] = 1
    evaluation_version: int = 1
    is_current: bool = True
    evaluated_at: Optional[datetime] = None
    created_at: Optional[datetime] = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: Optional[datetime] = Field(default_factory=lambda: datetime.now(timezone.utc))


class ComplianceSummaryCounts(BaseModel):
    """Aggregate counts for rule compliance states."""
    total: int = 0
    passed: int = 0
    failed: int = 0
    review: int = 0
    pending: int = 0
    not_applicable: int = 0
    blocked: int = 0
    mandatory_failures: int = 0
    critical_failures: int = 0


class BidComplianceSummaryResponse(BaseModel):
    """Complete bid compliance evaluation summary response."""
    model_config = ConfigDict(from_attributes=True)

    bid_id: uuid.UUID
    tender_id: uuid.UUID
    tender_number: Optional[str] = None
    bidder_name: Optional[str] = None
    compliance_evaluation_complete: bool = False
    counts: ComplianceSummaryCounts
    results: List[ComplianceResultItemResponse] = Field(default_factory=list)
    review_items: List[ComplianceReviewItemResponse] = Field(default_factory=list)
    evaluated_at: Optional[datetime] = None
